src/scheduler/mod.rs

```diff
@@ -27,12 +29,33 @@ pub struct Schedule {
     steps: Vec<ScheduleStep>,
 }
 
-#[derive(Clone, Debug, PartialEq, Eq)]
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
 enum ScheduleStep {
     Task(TaskId),
     Random,
 }
 
+/// A `ScheduleRecord` can be used to record both the step
+/// taken and all runnable tasks
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ScheduleRecord {
+    step: ScheduleStep,
+    runnable_tasks: HashSet<TaskId>
```

Comment:
Instead of a `HashSet`, could we just make this a `SmallVec<[TaskId; DEFAULT_INLINE_TASKS]>`? I think the invariant is that the `runnable` list we pass to schedulers is in a fixed order, so we can just compare them element by element to check if they're equal. (We need to document that invariant, though; I'll take care of that).